### ISSUE TITLE
QueryCursor#eachAsync uses O(N) stack frames.

### ISSUE THREAD

**Issue by Kurt-von-Laven** [Automattic/mongoose#4697]
I ran the script below with the command `node --stack-trace-limit=999999 eachAsyncBug.js 2> eachAsyncBug.log`, and it generated the following stack trace: https://gist.github.com/Kurt-von-Laven/7885fa35d0f540648c60e579e78b47ea. I am using mongoose 4.6.6 and mongodb version 2.6.11. The bug can be worked around by returning a promise (e.g., `return Promise.resolve();`) from `eachAsync`. I believe the fix is to avoid calling the callback directly here: https://github.com/Automattic/mongoose/blob/0c8b77637606f2f314af87ba043f2dccf190f857/lib/querycursor.js#L161. Instead it could be wrapped in a promise or `nextTick`, for example.

```
const mongoose = require('mongoose');
mongoose.Promise = global.Promise;

const NUM_RECORDS = 1000;

mongoose.connect('mongodb://localhost/myapp');

const Foo = mongoose.model('Foo', new mongoose.Schema({bar: String}));

Foo.remove().then(function() {
  let foos = [];
  for (let i = 0; i < NUM_RECORDS; i++) {
    foos.push({bar: 'baz'});
  }
  return Foo.create(foos);
}).then(function() {
  return Foo.find()
    .cursor()
    .eachAsync(function(foo) {
      // Don't return a promise.
    });
}).then(() => {
  return mongoose.connection.close();
});
```

**Comment by vkarpov15** [comment]
Thanks for the detailed repro, will be fixed in 4.6.8 :beers:

**Comment by Kurt-von-Laven** [comment]
Thanks for the quick fix!
